<a href="https://colab.research.google.com/github/anushayk70/Amazon-ML-Challenge/blob/main/notebook/01_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Environment Setup

In [ ]:
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


# 2. Dataset Discovery & Loading

In [ ]:
from pathlib import Path

base_path = Path("/content/drive/MyDrive/Amazon-ML-Challenge")

print("Files in the folder:\n")

for item in base_path.iterdir():
    print(item.name)

Files in the folder:

6ab10eb3b23ba_student_resource.zip


In [ ]:
import zipfile
from pathlib import Path

zip_path = Path(
    "/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip"
)

with zipfile.ZipFile(zip_path, "r") as z:
    files = z.namelist()

print("Number of files:", len(files))
print("\nFiles inside the ZIP:")

for f in files:
    print(f)

Number of files: 34

Files inside the ZIP:
student_resource/
__MACOSX/._student_resource
student_resource/Documentation_template.md
__MACOSX/student_resource/._Documentation_template.md
student_resource/.DS_Store
__MACOSX/student_resource/._.DS_Store
student_resource/dataset/
__MACOSX/student_resource/._dataset
student_resource/utils/
__MACOSX/student_resource/._utils
student_resource/README.md
__MACOSX/student_resource/._README.md
student_resource/dataset/.DS_Store
__MACOSX/student_resource/dataset/._.DS_Store
student_resource/dataset/test/
__MACOSX/student_resource/dataset/._test
student_resource/dataset/train/
__MACOSX/student_resource/dataset/._train
student_resource/utils/validate_submission.py
__MACOSX/student_resource/utils/._validate_submission.py
student_resource/dataset/test/test_source2.tsv
__MACOSX/student_resource/dataset/test/._test_source2.tsv
student_resource/dataset/test/test_source3.tsv
__MACOSX/student_resource/dataset/test/._test_source3.tsv
student_resource/dataset

In [ ]:
from pathlib import Path

print("Current folders/files:")

for path in Path("/content/drive").rglob("*"):
    if path.is_file() and path.suffix.lower() == ".zip":
        print(path)

Current folders/files:


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os

print(os.listdir("/content/drive/MyDrive"))

['Colab Notebooks', 'THE FLOWER.docx', 'worksheet-1.pdf', 'BRIGADE CCE LISTENING SKILLS', 'worksheet-1.gdoc', 'bioproject', 'Portions for TA1 - Std 9.docx', 'SIMULTANEOUS LINEAR EQUATIONS ix icse 2020-21.docx', 'TRIANGLES IX CBSE 2020-21.docx', 'TA1 TT Std 9 and 10 2020-21 (1).xlsx', 'TA1 TT Std 9 and 10 2020-21.xlsx', 'carbon tetrachlride amd methane.jpg', 'sodium chloride and calcium oxide.jpg', 'physics test.jpg', 'Classroom', 'jpg2pdf (28).pdf', 'jpg2pdf (29).pdf', 'physics ta 1 exams img1.jpg', 'STUDY OF THE FIRST ELEMENT - HYDROGEN.docx', '9th mix test.pdf', 'THE DIGESTIVE SYSTEM.gdoc', '9th mix test.gdoc', 'Portions for UT2 - Std 9 - 12.11.2020 (2).docx', 'page3.jpg', 'page11.jpg', 'page10.jpg', 'page7.jpg', 'page8.jpg', 'page12.jpg', 'page 16 (1).jpg', 'page5.jpg', 'page4.jpg', 'page13.jpg', 'page6.jpg', 'page 15 (1).jpg', 'page9.jpg', 'page1.jpg', 'page 17 (1).jpg', 'page 14 (1).jpg', 'page2.jpg', 'PHYSICS TUITIONS IX ICSE MIXED BATCH REFLECTION OF LIGHT WORKSHEET.pdf', 'STUDY

In [ ]:
import os

folder = "/content/drive/MyDrive/Amazon-ML-Challenge"

print("Files in Amazon-ML-Challenge:")
for file in os.listdir(folder):
    print(repr(file))

Files in Amazon-ML-Challenge:
'6ab10eb3b23ba_student_resource.zip'


In [ ]:
zip_path = os.path.join(folder, os.listdir(folder)[0])

print("ZIP path:")
print(zip_path)

ZIP path:
/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip


In [ ]:
import zipfile
import os

folder = "/content/drive/MyDrive/Amazon-ML-Challenge"

# Automatically find the ZIP
zip_files = [
    f for f in os.listdir(folder)
    if f.lower().endswith(".zip")
]

zip_path = os.path.join(folder, zip_files[0])

print("Using ZIP:")
print(zip_path)

# Read README directly from ZIP
with zipfile.ZipFile(zip_path, "r") as z:
    readme = z.read("student_resource/README.md").decode("utf-8")

print("\n" + "="*70)
print("ORGANIZER README")
print("="*70 + "\n")
print(readme)

Using ZIP:
/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip

ORGANIZER README

# ML Challenge 2026 Problem Statement

## Business Entity Resolution Challenge

In large-scale commercial platforms, business identity data arrives from multiple independent sources — each contributing partial, noisy fragments of information about the same real-world entities. These fragments share no common identifiers, and the challenge of determining which records refer to the same business is known as Entity Resolution (ER). Your challenge is to build an ML solution that, given business records from 3 independent data sources with noisy and inconsistent fields, determines which records across sources refer to the same real-world business entity.

Source 1 is the deduplicated reference source. Your task is to find all matching records from Source 2 and Source 3 for each Source 1 entity. A Source 1 entity may match zero, one, or many records from Source 2 and Source 3.

### Fil

In [ ]:
import zipfile
import pandas as pd

zip_path = "/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip"

with zipfile.ZipFile(zip_path, "r") as z:

    train_s1 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source1.tsv"),
        sep="\t"
    )

    train_s2 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source2.tsv"),
        sep="\t"
    )

    train_s3 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source3.tsv"),
        sep="\t"
    )

    ground_truth = pd.read_csv(
        z.open("student_resource/dataset/train/train_ground_truth.tsv"),
        sep="\t"
    )

print("Source 1:", train_s1.shape)
print("Source 2:", train_s2.shape)
print("Source 3:", train_s3.shape)
print("Ground Truth:", ground_truth.shape)

Source 1: (2206821, 4)
Source 2: (5034616, 4)
Source 3: (5285603, 4)
Ground Truth: (2206821, 2)


# 3. Dataset Overview

In [ ]:
print("SOURCE 1")
display(train_s1.head())

print("\nSOURCE 2")
display(train_s2.head())

print("\nSOURCE 3")
display(train_s3.head())

print("\nGROUND TRUTH")
display(ground_truth.head())

SOURCE 1


,entity_id,business_name,business_address,country
0,S1-925783039,Orelee's Barbershop,"1795 Westchester Drive, High Point, NC",US
1,S1-773889195,Prime Money,"17560 Ellis Road, Tahlequah, OK",US
2,S1-377745466,B+ Retail Inc,"1712 Montebello Avenue, Phoenix, AZ",US
3,S1-133037285,Christ Chapel,"2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD",US
4,S1-755362802,Prabhav Business Center,"797, Lake Town Block A, Kolkata, Howrah, West ...",India



SOURCE 2


,entity_id,business_name,business_address,country
0,S2-166376419,राम मार्केटिंग प्राइवेट लिमिटेड,"KH NO. -570/13, NEW DELHI, WEST DELHI, Delhi",India
1,S2-764573417,-- Holloway Peak Inc Seafood,"105 ELM ST, MORGANTON, NC",US
2,S2-639257739,आदित्य प्रॉपर्टीज एलएलपी,"G-3/571, GULMOHAR COLONY, BHOPAL, Madhya Pradesh",India
3,S2-163963287,Summit Inc,"GREENSBORO, NC, 19 1/2 STARDUST TRAIL",US
4,S2-49942811,Delta Tetlecommunication Inc,"914 PIERPONT AVE, CLEVELAND, OH",US



SOURCE 3


,entity_id,business_name,business_address,country
0,S3-202863386,wilfordhancock.com,"Mack Rd, Haltom City, Texas",US
1,S3-859268022,International South Consultants Private Ltd,NaN,India
2,S3-22467283,LLC Moncada Léarning Center,"5780 Fawn Ct, Fort Worth, Texas",US
3,S3-671162755,Moyna's Coffee,"1 Ivanhoe Ave, PO Box 6009, Cincinnati, Ohio",US
4,S3-960981775,Pvt. EFS Print Ventures Ltd.,"Door No 183, 41St Cross, 22Nd Main 9Th Block J...",India



GROUND TRUTH


,source1_entity_id,matched_entity_ids
0,S1-965667,"S2-681193310,S2-743505751,S3-775321672,S3-1129..."
1,S1-55344266,"S2-249013014,S2-197070651,S3-478195123,S3-3843..."
2,S1-343815751,"S2-790675320,S2-479876582,S3-878454467"
3,S1-656753428,"S2-153058913,S2-24659151,S3-679606215"
4,S1-102811957,"S2-478959098,S2-553508714,S2-625774905,S3-7280..."


In [ ]:
import os
import glob

# Find all ZIP files inside the Amazon-ML-Challenge folder
zip_files = glob.glob(
    "/content/drive/MyDrive/Amazon-ML-Challenge/*.zip"
)

print("ZIP files found:")
for f in zip_files:
    print(f)

ZIP files found:
/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip


In [ ]:
import zipfile
import os

# Use the ZIP file we found
zip_path = zip_files[0]

# Extraction location
extract_path = "/content/amazon_ml_data"

# Extract
with zipfile.ZipFile(zip_path, "r") as z:
    z.extractall(extract_path)

print("Dataset extracted successfully!")
print("\nTop-level contents:")
print(os.listdir(extract_path))

Dataset extracted successfully!

Top-level contents:
['__MACOSX', 'student_resource']


In [ ]:
import os

for root, dirs, files in os.walk(extract_path):
    level = root.replace(extract_path, "").count(os.sep)
    indent = "    " * level
    print(f"{indent}{os.path.basename(root)}/")

    for file in files:
        print(f"{indent}    {file}")

amazon_ml_data/
    __MACOSX/
        ._student_resource
        student_resource/
            ._README.md
            ._utils
            ._dataset
            ._.DS_Store
            ._Documentation_template.md
            dataset/
                ._train
                ._.DS_Store
                ._test
                test/
                    ._test_source1.tsv
                    ._test_source3.tsv
                    ._test_source2.tsv
                train/
                    ._train_source3.tsv
                    ._train_ground_truth.tsv
                    ._train_source1.tsv
                    ._train_source2.tsv
            utils/
                ._validate_submission.py
    student_resource/
        .DS_Store
        README.md
        Documentation_template.md
        dataset/
            .DS_Store
            test/
                test_source3.tsv
                test_source1.tsv
                test_source2.tsv
            train/
                train_ground_truth.ts

In [ ]:
import pandas as pd

train_path = "/content/amazon_ml_data/student_resource/dataset/train"

train_s1 = pd.read_csv(
    f"{train_path}/train_source1.tsv",
    sep="\t"
)

train_s2 = pd.read_csv(
    f"{train_path}/train_source2.tsv",
    sep="\t"
)

train_s3 = pd.read_csv(
    f"{train_path}/train_source3.tsv",
    sep="\t"
)

ground_truth = pd.read_csv(
    f"{train_path}/train_ground_truth.tsv",
    sep="\t"
)

print("Source 1 shape:", train_s1.shape)
print("Source 2 shape:", train_s2.shape)
print("Source 3 shape:", train_s3.shape)
print("Ground Truth shape:", ground_truth.shape)

Source 1 shape: (2206821, 4)
Source 2 shape: (5034616, 4)
Source 3 shape: (5285603, 4)
Ground Truth shape: (2206821, 2)


In [ ]:
print("SOURCE 1 COLUMNS")
print(train_s1.columns.tolist())

print("\nSOURCE 2 COLUMNS")
print(train_s2.columns.tolist())

print("\nSOURCE 3 COLUMNS")
print(train_s3.columns.tolist())

print("\nGROUND TRUTH COLUMNS")
print(ground_truth.columns.tolist())

SOURCE 1 COLUMNS
['entity_id', 'business_name', 'business_address', 'country']

SOURCE 2 COLUMNS
['entity_id', 'business_name', 'business_address', 'country']

SOURCE 3 COLUMNS
['entity_id', 'business_name', 'business_address', 'country']

GROUND TRUTH COLUMNS
['source1_entity_id', 'matched_entity_ids']


In [ ]:
print("========== SOURCE 1 ==========")
display(train_s1.head(10))

print("========== SOURCE 2 ==========")
display(train_s2.head(10))

print("========== SOURCE 3 ==========")
display(train_s3.head(10))

print("========== GROUND TRUTH ==========")
display(ground_truth.head(10))

========== SOURCE 1 ==========


,entity_id,business_name,business_address,country
0,S1-925783039,Orelee's Barbershop,"1795 Westchester Drive, High Point, NC",US
1,S1-773889195,Prime Money,"17560 Ellis Road, Tahlequah, OK",US
2,S1-377745466,B+ Retail Inc,"1712 Montebello Avenue, Phoenix, AZ",US
3,S1-133037285,Christ Chapel,"2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD",US
4,S1-755362802,Prabhav Business Center,"797, Lake Town Block A, Kolkata, Howrah, West ...",India
5,S1-851869949,Custom Wealth Services LLC,"OH, Columbus, 5559 Orville Avenue",US
6,S1-785847572,Consulting Nyasa Nursing Private Limited,"2505, Tower 1, Oakwood, Runwal Greens, Mulund ...",India
7,S1-27541239,Nexus Anchor Rain,"1111 Church Street, Unit 2007, Nashville, TN",US
8,S1-629417405,Moore Bitwise Inc,"337 Oakland Avenue, Michigan City, IN",US
9,S1-22305073,Dermatology Green Medicine,"294 Meadowcreek Drive, Unit Unit 2, Village Of...",US


========== SOURCE 2 ==========


,entity_id,business_name,business_address,country
0,S2-166376419,राम मार्केटिंग प्राइवेट लिमिटेड,"KH NO. -570/13, NEW DELHI, WEST DELHI, Delhi",India
1,S2-764573417,-- Holloway Peak Inc Seafood,"105 ELM ST, MORGANTON, NC",US
2,S2-639257739,आदित्य प्रॉपर्टीज एलएलपी,"G-3/571, GULMOHAR COLONY, BHOPAL, Madhya Pradesh",India
3,S2-163963287,Summit Inc,"GREENSBORO, NC, 19 1/2 STARDUST TRAIL",US
4,S2-49942811,Delta Tetlecommunication Inc,"914 PIERPONT AVE, CLEVELAND, OH",US
5,S2-138046867,Lee and Lawson,"1702 Pine Avenue, CITY OF MENOMONIE, WI",US
6,S2-584977605,SHIVSHAKTI VIDYALAYA VIDYALAYA OVERSEAS CORPOR...,"H.NO 204 C ROAD HOSHIARPUR, PUNJAB, Punjab",India
7,S2-277444929,Shree Infracon Private Ltd,"63/2275/7, ALHIND TOWER, FIRST FLOOR, JAFFERKH...",India
8,S2-721031885,Chavira Platinum Chimera LLC,"282 SAXONY DRIVE, FTT MITCHELL, KY",US
9,S2-508602797,FOUNDATION EXCEL AGENCY PRIVATE LIMITED,"HN 753 E-1, BHARAT NAGAR, 104/1/1 ERANDWANE, M...",India


========== SOURCE 3 ==========


,entity_id,business_name,business_address,country
0,S3-202863386,wilfordhancock.com,"Mack Rd, Haltom City, Texas",US
1,S3-859268022,International South Consultants Private Ltd,NaN,India
2,S3-22467283,LLC Moncada Léarning Center,"5780 Fawn Ct, Fort Worth, Texas",US
3,S3-671162755,Moyna's Coffee,"1 Ivanhoe Ave, PO Box 6009, Cincinnati, Ohio",US
4,S3-960981775,Pvt. EFS Print Ventures Ltd.,"Door No 183, 41St Cross, 22Nd Main 9Th Block J...",India
5,S3-578159284,LLC Hernandez Colonial Redwood,"2260- Housecreek Trail, Unit 407, Raleigh, Nor...",US
6,S3-121412624,Classic Equity Partners Group,"6885 Catalpa Bluff Ln, PO Box 5799, Dickinson,...",US
7,S3-249416830,Ectolumdrex dba X+ Madison Inc,"S03575 Cty Tk M, Town Of Buffalo, WI",US
8,S3-107644605,Animal Hanisch Hospirlg,"##8 Willow Oak Lane, Fl. 0, Saint Louis, Missouri",US
9,S3-160217003,Gomez Optimal,"343 Hempstead 161, Hope, Arkansas",US


========== GROUND TRUTH ==========


,source1_entity_id,matched_entity_ids
0,S1-965667,"S2-681193310,S2-743505751,S3-775321672,S3-1129..."
1,S1-55344266,"S2-249013014,S2-197070651,S3-478195123,S3-3843..."
2,S1-343815751,"S2-790675320,S2-479876582,S3-878454467"
3,S1-656753428,"S2-153058913,S2-24659151,S3-679606215"
4,S1-102811957,"S2-478959098,S2-553508714,S2-625774905,S3-7280..."
5,S1-18727616,"S2-755677256,S3-187831601,S3-641489370,S3-4762..."
6,S1-318373630,"S2-660036492,S3-804600254"
7,S1-86989137,"S3-274817120,S3-312496301"
8,S1-29845983,"S2-648035184,S3-588502663"
9,S1-789009573,"S2-383871912,S3-74481402,S3-576451439"


# 4. Missing Value Analysis

In [ ]:
import os
import glob

# Find all ZIP files inside the Amazon-ML-Challenge folder
zip_files = glob.glob(
    "/content/drive/MyDrive/Amazon-ML-Challenge/*.zip"
)

print("ZIP files found:")
for f in zip_files:
    print(f)

ZIP files found:
/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip


In [ ]:
print("SOURCE 1 — MISSING VALUES")
display(train_s1.isnull().sum())

print("\nSOURCE 2 — MISSING VALUES")
display(train_s2.isnull().sum())

print("\nSOURCE 3 — MISSING VALUES")
display(train_s3.isnull().sum())

print("\nGROUND TRUTH — MISSING VALUES")
display(ground_truth.isnull().sum())

SOURCE 1 — MISSING VALUES


,0
entity_id,0
business_name,0
business_address,0
country,0



SOURCE 2 — MISSING VALUES


,0
entity_id,0
business_name,2
business_address,168967
country,0



SOURCE 3 — MISSING VALUES


,0
entity_id,0
business_name,13
business_address,175916
country,0



GROUND TRUTH — MISSING VALUES


,0
source1_entity_id,0
matched_entity_ids,123247


In [ ]:
datasets = {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3,
    "Ground Truth": ground_truth
}

for name, df in datasets.items():
    print(f"\n========== {name} ==========")

    missing_count = df.isnull().sum()
    missing_percentage = (df.isnull().mean() * 100).round(2)

    missing_summary = pd.DataFrame({
        "Missing Count": missing_count,
        "Missing Percentage": missing_percentage
    })

    display(missing_summary)


========== Source 1 ==========


,Missing Count,Missing Percentage
entity_id,0,0.0
business_name,0,0.0
business_address,0,0.0
country,0,0.0



========== Source 2 ==========


,Missing Count,Missing Percentage
entity_id,0,0.00
business_name,2,0.00
business_address,168967,3.36
country,0,0.00



========== Source 3 ==========


,Missing Count,Missing Percentage
entity_id,0,0.00
business_name,13,0.00
business_address,175916,3.33
country,0,0.00



========== Ground Truth ==========


,Missing Count,Missing Percentage
source1_entity_id,0,0.00
matched_entity_ids,123247,5.58


# 5. Duplicate Analysis

In [ ]:
datasets = {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}

for name, df in datasets.items():
    print(f"\n{'='*50}")
    print(name)
    print(f"{'='*50}")

    print("Duplicate complete rows:", df.duplicated().sum())
    print("Duplicate entity IDs:", df["entity_id"].duplicated().sum())
    print("Duplicate business names:", df["business_name"].duplicated().sum())
    print("Duplicate business addresses:", df["business_address"].duplicated().sum())


Source 1
Duplicate complete rows: 0
Duplicate entity IDs: 0
Duplicate business names: 667592
Duplicate business addresses: 76215

Source 2
Duplicate complete rows: 0
Duplicate entity IDs: 0
Duplicate business names: 632607
Duplicate business addresses: 697354

Source 3
Duplicate complete rows: 0
Duplicate entity IDs: 0
Duplicate business names: 633994
Duplicate business addresses: 652838


# 6. Uniqueness Analysis

In [ ]:
datasets = {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}

for name, df in datasets.items():
    total = len(df)

    unique_names = df["business_name"].nunique(dropna=True)
    unique_addresses = df["business_address"].nunique(dropna=True)

    print(f"\n{'='*50}")
    print(name)
    print(f"{'='*50}")

    print("Total records:", total)
    print("Unique business names:", unique_names)
    print("Unique business addresses:", unique_addresses)

    print(
        "Name uniqueness percentage:",
        round(unique_names / total * 100, 2), "%"
    )

    print(
        "Address uniqueness percentage:",
        round(unique_addresses / total * 100, 2), "%"
    )


Source 1
Total records: 2206821
Unique business names: 1539229
Unique business addresses: 2130606
Name uniqueness percentage: 69.75 %
Address uniqueness percentage: 96.55 %

Source 2
Total records: 5034616
Unique business names: 4402008
Unique business addresses: 4337261
Name uniqueness percentage: 87.43 %
Address uniqueness percentage: 86.15 %

Source 3
Total records: 5285603
Unique business names: 4651608
Unique business addresses: 4632764
Name uniqueness percentage: 88.01 %
Address uniqueness percentage: 87.65 %


# 7. Country Distribution

In [ ]:
datasets = {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}

for name, df in datasets.items():
    print(f"\n{'='*60}")
    print(f"{name} — COUNTRY DISTRIBUTION")
    print(f"{'='*60}")

    country_counts = df["country"].value_counts()

    country_summary = pd.DataFrame({
        "Count": country_counts,
        "Percentage": (country_counts / len(df) * 100).round(2)
    })

    display(country_summary)


Source 1 — COUNTRY DISTRIBUTION


,Count,Percentage
country,,
US,1323633,59.98
India,883188,40.02



Source 2 — COUNTRY DISTRIBUTION


,Count,Percentage
country,,
US,3016817,59.92
India,2017799,40.08



Source 3 — COUNTRY DISTRIBUTION


,Count,Percentage
country,,
US,3170056,59.98
India,2115547,40.02


# 8. Entity ID Analysis

In [ ]:
for name, df in {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}.items():

    print(f"\n{'='*60}")
    print(f"{name} — ENTITY ID ANALYSIS")
    print(f"{'='*60}")

    print("Data type:", df["entity_id"].dtype)
    print("Number of unique IDs:", df["entity_id"].nunique())
    print("Number of records:", len(df))

    print("\nFirst 10 IDs:")
    print(df["entity_id"].head(10).tolist())

    print("\nLast 10 IDs:")
    print(df["entity_id"].tail(10).tolist())


Source 1 — ENTITY ID ANALYSIS
Data type: object
Number of unique IDs: 2206821
Number of records: 2206821

First 10 IDs:
['S1-925783039', 'S1-773889195', 'S1-377745466', 'S1-133037285', 'S1-755362802', 'S1-851869949', 'S1-785847572', 'S1-27541239', 'S1-629417405', 'S1-22305073']

Last 10 IDs:
['S1-178574324', 'S1-668548746', 'S1-395658450', 'S1-199645528', 'S1-419336635', 'S1-479751630', 'S1-392755726', 'S1-716147977', 'S1-770136229', 'S1-252639948']

Source 2 — ENTITY ID ANALYSIS
Data type: object
Number of unique IDs: 5034616
Number of records: 5034616

First 10 IDs:
['S2-166376419', 'S2-764573417', 'S2-639257739', 'S2-163963287', 'S2-49942811', 'S2-138046867', 'S2-584977605', 'S2-277444929', 'S2-721031885', 'S2-508602797']

Last 10 IDs:
['S2-15536678', 'S2-794164146', 'S2-896466345', 'S2-345965751', 'S2-330974843', 'S2-147379319', 'S2-828699622', 'S2-726846711', 'S2-115058866', 'S2-955929147']

Source 3 — ENTITY ID ANALYSIS
Data type: object
Number of unique IDs: 5285603
Number of r

# 9. Business Name Analysis

## 9.1 Business Name Length

In [ ]:
datasets = {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}

for name, df in datasets.items():

    name_lengths = df["business_name"].dropna().astype(str).str.len()

    print(f"\n{'='*60}")
    print(f"{name} — BUSINESS NAME LENGTH")
    print(f"{'='*60}")

    print("Minimum:", name_lengths.min())
    print("Maximum:", name_lengths.max())
    print("Mean:", round(name_lengths.mean(), 2))
    print("Median:", name_lengths.median())
    print("25th percentile:", name_lengths.quantile(0.25))
    print("75th percentile:", name_lengths.quantile(0.75))
    print("95th percentile:", name_lengths.quantile(0.95))


Source 1 — BUSINESS NAME LENGTH
Minimum: 3
Maximum: 105
Mean: 24.03
Median: 24.0
25th percentile: 18.0
75th percentile: 30.0
95th percentile: 37.0

Source 2 — BUSINESS NAME LENGTH
Minimum: 2
Maximum: 104
Mean: 25.1
Median: 25.0
25th percentile: 19.0
75th percentile: 31.0
95th percentile: 40.0

Source 3 — BUSINESS NAME LENGTH
Minimum: 2
Maximum: 123
Mean: 25.2
Median: 25.0
25th percentile: 18.0
75th percentile: 31.0
95th percentile: 42.0


## 9.2 Top 20 Business Names

In [ ]:
for name, df in {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}.items():

    print(f"\n{'='*60}")
    print(f"{name} — TOP 20 BUSINESS NAMES")
    print(f"{'='*60}")

    top_names = df["business_name"].value_counts(dropna=False).head(20)

    display(
        top_names.rename("Count").to_frame()
    )


Source 1 — TOP 20 BUSINESS NAMES


,Count
business_name,
Primary Care Group,253
Ear Nose & Throat Group,251
Pediatric Group,222
Womens Health Group,220
Physical Therapy Group,218
Pediatric Dental Group,216
Behavioral Health Group,215
Chiropractic Group,209
Orthopedic Group,208



Source 2 — TOP 20 BUSINESS NAMES


,Count
business_name,
Primary Care,320
Physical Therapy,307
Womens Health,297
Urgent Care,297
Behavioral Health,296
Internal Medicine,288
Pediatric Dentistry,274
earnosethroat.com,269
Pediatric Dental,268



Source 3 — TOP 20 BUSINESS NAMES


,Count
business_name,
Primary Care,421
Physical Therapy,399
Pediatric Dental,393
Womens Health,379
Urgent Care,377
Pediatric Dentistry,361
Behavioral Health,330
Pediatric,321
Internal Medicine,319


In [ ]:
for name, df in {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}.items():

    s = df["business_name"].dropna().astype(str)

    print(f"\n{'='*60}")
    print(f"{name} — BUSINESS NAME CHARACTER COMPOSITION")
    print(f"{'='*60}")

    print("Names containing digits:",
          s.str.contains(r"\d", regex=True).sum())

    print("Names containing punctuation/special characters:",
          s.str.contains(r"[^A-Za-z0-9\s]", regex=True).sum())

    print("Names containing spaces:",
          s.str.contains(r"\s", regex=True).sum())

    print("Names containing only letters:",
          s.str.fullmatch(r"[A-Za-z]+", regex=True).sum())

    print("Names containing only letters and spaces:",
          s.str.fullmatch(r"[A-Za-z\s]+", regex=True).sum())


Source 1 — BUSINESS NAME CHARACTER COMPOSITION
Names containing digits: 35585
Names containing punctuation/special characters: 461669
Names containing spaces: 2191404


TypeError: StringMethods.fullmatch() got an unexpected keyword argument 'regex'

## 9.3 Business Name Character Composition

In [ ]:
for name, df in {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}.items():

    s = df["business_name"].dropna().astype(str)

    print(f"\n{'='*60}")
    print(f"{name} — BUSINESS NAME CHARACTER COMPOSITION")
    print(f"{'='*60}")

    print("Contains digits:",
          s.str.contains(r"\d", regex=True).sum())

    print("Contains punctuation/special characters:",
          s.str.contains(r"[^A-Za-z0-9\s]", regex=True).sum())

    print("Contains spaces:",
          s.str.contains(r"\s", regex=True).sum())

    print("Only letters:",
          s.str.fullmatch(r"[A-Za-z]+").sum())

    print("Only letters + spaces:",
          s.str.fullmatch(r"[A-Za-z\s]+").sum())


Source 1 — BUSINESS NAME CHARACTER COMPOSITION
Contains digits: 35585
Contains punctuation/special characters: 461669
Contains spaces: 2191404
Only letters: 14598
Only letters + spaces: 1718343

Source 2 — BUSINESS NAME CHARACTER COMPOSITION
Contains digits: 255636
Contains punctuation/special characters: 2356371
Contains spaces: 4722467
Only letters: 69503
Only letters + spaces: 2555437

Source 3 — BUSINESS NAME CHARACTER COMPOSITION
Contains digits: 270847
Contains punctuation/special characters: 2347044
Contains spaces: 4889076
Only letters: 139059
Only letters + spaces: 2809037


## 9.4 Empty / Whitespace Business Names

In [ ]:
import pandas as pd

datasets = {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}

# ============================================================
# 1. WHITESPACE / EMPTY STRING CHECK
# ============================================================

print("\n" + "="*70)
print("EMPTY / WHITESPACE BUSINESS NAMES")
print("="*70)

for name, df in datasets.items():
    s = df["business_name"].dropna().astype(str)

    empty = (s.str.strip() == "").sum()

    print(f"{name}: {empty}")


# ============================================================
# 2. CASE VARIATION
# ============================================================

print("\n" + "="*70)
print("BUSINESS NAMES AFFECTED BY CASE")
print("="*70)

for name, df in datasets.items():
    s = df["business_name"].dropna().astype(str)

    case_duplicates = (
        s.str.lower().nunique()
    )

    original_unique = s.nunique()

    print(f"{name}")
    print("Original unique names:", original_unique)
    print("Lowercase unique names:", case_duplicates)
    print("Potential case-collapsed names:",
          original_unique - case_duplicates)


# ============================================================
# 3. ADDRESS LENGTH
# ============================================================

print("\n" + "="*70)
print("BUSINESS ADDRESS LENGTH")
print("="*70)

for name, df in datasets.items():

    s = df["business_address"].dropna().astype(str)
    lengths = s.str.len()

    print(f"\n{name}")
    print("Minimum:", lengths.min())
    print("Maximum:", lengths.max())
    print("Mean:", round(lengths.mean(), 2))
    print("Median:", lengths.median())
    print("95th percentile:", lengths.quantile(0.95))


# ============================================================
# 4. GROUND TRUTH MATCH COUNTS
# ============================================================

print("\n" + "="*70)
print("GROUND TRUTH MATCH COUNTS")
print("="*70)

gt = ground_truth.copy()

def count_matches(value):
    if pd.isna(value):
        return 0

    value = str(value).strip()

    if value == "":
        return 0

    # Handle common separators used in match lists
    if "|" in value:
        return len([x for x in value.split("|") if x.strip()])

    if "," in value:
        return len([x for x in value.split(",") if x.strip()])

    return 1


gt["match_count"] = gt["matched_entity_ids"].apply(count_matches)

print(gt["match_count"].value_counts().sort_index())


# ============================================================
# 5. GROUND TRUTH SUMMARY
# ============================================================

print("\n" + "="*70)
print("GROUND TRUTH SUMMARY")
print("="*70)

print("Total Source 1 entities:", len(gt))
print("Zero-match entities:", (gt["match_count"] == 0).sum())
print("One-match entities:", (gt["match_count"] == 1).sum())
print("Multiple-match entities:", (gt["match_count"] > 1).sum())


EMPTY / WHITESPACE BUSINESS NAMES
Source 1: 0
Source 2: 0
Source 3: 0

BUSINESS NAMES AFFECTED BY CASE
Source 1
Original unique names: 1539229
Lowercase unique names: 1538804
Potential case-collapsed names: 425
Source 2
Original unique names: 4402008
Lowercase unique names: 4287470
Potential case-collapsed names: 114538
Source 3
Original unique names: 4651608
Lowercase unique names: 4579083
Potential case-collapsed names: 72525

BUSINESS ADDRESS LENGTH

Source 1
Minimum: 11
Maximum: 256
Mean: 52.07
Median: 41.0
95th percentile: 103.0

Source 2
Minimum: 8
Maximum: 249
Mean: 47.83
Median: 37.0
95th percentile: 97.0

Source 3
Minimum: 2
Maximum: 240
Mean: 48.32
Median: 42.0
95th percentile: 92.0

GROUND TRUTH MATCH COUNTS
match_count
0     123247
1     119157
2     375212
3     530841
4     484115
5     321957
6     164868
7      63968
8      18680
9       4205
10       534
11        37
Name: count, dtype: int64

GROUND TRUTH SUMMARY
Total Source 1 entities: 2206821
Zero-match entities: 

In [ ]:
print(ground_truth[["source1_entity_id", "matched_entity_ids"]].head(10).to_string())

  source1_entity_id                                                             matched_entity_ids
0         S1-965667                S2-681193310,S2-743505751,S3-775321672,S3-11291185,S3-860443364
1       S1-55344266                            S2-249013014,S2-197070651,S3-478195123,S3-384364074
2      S1-343815751                                         S2-790675320,S2-479876582,S3-878454467
3      S1-656753428                                          S2-153058913,S2-24659151,S3-679606215
4      S1-102811957  S2-478959098,S2-553508714,S2-625774905,S3-728090388,S3-928796641,S3-449308785
5       S1-18727616                            S2-755677256,S3-187831601,S3-641489370,S3-476250621
6      S1-318373630                                                      S2-660036492,S3-804600254
7       S1-86989137                                                      S3-274817120,S3-312496301
8       S1-29845983                                                      S2-648035184,S3-588502663
9      S1-

##9.5 Ground-truth mapping

In [ ]:
# Create ground-truth mapping

ground_truth_map = {}

for _, row in ground_truth.iterrows():

    source1_id = row["source1_entity_id"]
    matched_ids = row["matched_entity_ids"]

    if pd.isna(matched_ids) or str(matched_ids).strip() == "":
        ground_truth_map[source1_id] = set()
    else:
        ground_truth_map[source1_id] = set(
            x.strip()
            for x in str(matched_ids).split(",")
            if x.strip()
        )

print("Number of Source 1 entities:", len(ground_truth_map))

# Example
for source1_id in list(ground_truth_map.keys())[:5]:
    print(source1_id, "→", ground_truth_map[source1_id])

Number of Source 1 entities: 2206821
S1-965667 → {'S3-775321672', 'S3-860443364', 'S3-11291185', 'S2-681193310', 'S2-743505751'}
S1-55344266 → {'S2-197070651', 'S2-249013014', 'S3-384364074', 'S3-478195123'}
S1-343815751 → {'S2-790675320', 'S3-878454467', 'S2-479876582'}
S1-656753428 → {'S2-153058913', 'S3-679606215', 'S2-24659151'}
S1-102811957 → {'S3-928796641', 'S3-449308785', 'S3-728090388', 'S2-625774905', 'S2-478959098', 'S2-553508714'}


In [ ]:
source23 = pd.concat(
    [
        train_s2.assign(source="S2"),
        train_s3.assign(source="S3")
    ],
    ignore_index=True
)

print("Source 2 records:", len(train_s2))
print("Source 3 records:", len(train_s3))
print("Combined Source 2 + Source 3:", len(source23))

print("\nColumns:")
print(source23.columns.tolist())

Source 2 records: 5034616
Source 3 records: 5285603
Combined Source 2 + Source 3: 10320219

Columns:
['entity_id', 'business_name', 'business_address', 'country', 'source']


In [ ]:
import re

def normalize_text(text):
    if pd.isna(text):
        return ""

    text = str(text).lower()

    # Replace punctuation/special characters with spaces
    text = re.sub(r"[^a-z0-9\s]", " ", text)

    # Normalize whitespace
    text = re.sub(r"\s+", " ", text).strip()

    return text

In [ ]:
examples = [
    "ABC PRIVATE LIMITED",
    "ABC Pvt. Ltd.",
    "  ABC   Pharmacy  ",
    "Ear Nose & Throat Group"
]

for x in examples:
    print(f"{x}  →  {normalize_text(x)}")

ABC PRIVATE LIMITED  →  abc private limited
ABC Pvt. Ltd.  →  abc pvt ltd
  ABC   Pharmacy    →  abc pharmacy
Ear Nose & Throat Group  →  ear nose throat group


In [ ]:
# Create normalized text columns

for df in [train_s1, train_s2, train_s3]:
    df["name_normalized"] = df["business_name"].apply(normalize_text)
    df["address_normalized"] = df["business_address"].apply(normalize_text)

print("Source 1:")
print(train_s1[[
    "business_name",
    "name_normalized",
    "business_address",
    "address_normalized"
]].head(5).to_string(index=False))

print("\nSource 2:")
print(train_s2[[
    "business_name",
    "name_normalized",
    "business_address",
    "address_normalized"
]].head(5).to_string(index=False))

Source 1:
          business_name         name_normalized                                     business_address                               address_normalized
    Orelee's Barbershop     orelee s barbershop               1795 Westchester Drive, High Point, NC             1795 westchester drive high point nc
            Prime Money             prime money                      17560 Ellis Road, Tahlequah, OK                    17560 ellis road tahlequah ok
          B+ Retail Inc            b retail inc                  1712 Montebello Avenue, Phoenix, AZ                1712 montebello avenue phoenix az
          Christ Chapel           christ chapel    2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD   2100 cameron drive unit apartment g dundalk md
Prabhav Business Center prabhav business center 797, Lake Town Block A, Kolkata, Howrah, West Bengal 797 lake town block a kolkata howrah west bengal

Source 2:
                  business_name              name_normalized                   

In [ ]:
import re
import unicodedata

def normalize_text(text):
    if pd.isna(text):
        return ""

    text = str(text).lower()

    # Normalize Unicode representation
    text = unicodedata.normalize("NFKC", text)

    # Remove punctuation while PRESERVING Unicode letters/numbers
    text = "".join(
        char if char.isalnum() or char.isspace() else " "
        for char in text
    )

    # Normalize whitespace
    text = re.sub(r"\s+", " ", text).strip()

    return text

In [ ]:
# Re-create normalized columns using the corrected function

for df in [train_s1, train_s2, train_s3]:
    df["name_normalized"] = df["business_name"].apply(normalize_text)
    df["address_normalized"] = df["business_address"].apply(normalize_text)

print(train_s2[[
    "business_name",
    "name_normalized"
]].head(10).to_string(index=False))

KeyboardInterrupt: 

In [ ]:
import re
import unicodedata

def normalize_series(s):
    s = s.fillna("").astype("string")
    s = s.str.lower()
    s = s.str.normalize("NFKC")

    # Keep Unicode letters/numbers and whitespace
    s = s.str.replace(r"[^\w\s]", " ", regex=True)

    # Remove underscore separately
    s = s.str.replace("_", " ", regex=False)

    # Normalize whitespace
    s = s.str.replace(r"\s+", " ", regex=True).str.strip()

    return s

In [ ]:
import re
import unicodedata

def normalize_series(s):
    s = s.fillna("").astype("string")
    s = s.str.lower()
    s = s.str.normalize("NFKC")

    # Keep Unicode letters/numbers and whitespace
    s = s.str.replace(r"[^\w\s]", " ", regex=True)

    # Remove underscore separately
    s = s.str.replace("_", " ", regex=False)

    # Normalize whitespace
    s = s.str.replace(r"\s+", " ", regex=True).str.strip()

    return s

In [ ]:
print("Source 1 normalized columns:")
print(train_s1[["business_name", "name_normalized"]].head(5).to_string(index=False))

print("\nSource 2 normalized columns:")
print(train_s2[["business_name", "name_normalized"]].head(10).to_string(index=False))

Source 1 normalized columns:
          business_name         name_normalized
    Orelee's Barbershop     orelee s barbershop
            Prime Money             prime money
          B+ Retail Inc            b retail inc
          Christ Chapel           christ chapel
Prabhav Business Center prabhav business center

Source 2 normalized columns:
                                                           business_name                                                        name_normalized
                                         राम मार्केटिंग प्राइवेट लिमिटेड                                         र म म र क ट ग प र इव ट ल म ट ड
                                            -- Holloway Peak Inc Seafood                                              holloway peak inc seafood
                                                आदित्य प्रॉपर्टीज एलएलपी                                                आद त य प र पर ट ज एलएलप
                                                              Summit Inc     

# 10. Business Address Analysis

In [ ]:
# STEP 6 — BLOCKING KEYS

for df in [train_s1, train_s2, train_s3]:
    df["name_block"] = (
        df["name_normalized"]
        .fillna("")
        .str.replace(" ", "", regex=False)
        .str[:10]
    )

print("Source 1 blocking keys:")
print(
    train_s1[["business_name", "country", "name_block"]]
    .head(10)
    .to_string(index=False)
)

print("\nNumber of unique Source 1 blocks:",
      train_s1["name_block"].nunique())

print("Number of unique Source 2 blocks:",
      train_s2["name_block"].nunique())

print("Number of unique Source 3 blocks:",
      train_s3["name_block"].nunique())

Source 1 blocking keys:
                           business_name country name_block
                     Orelee's Barbershop      US oreleesbar
                             Prime Money      US primemoney
                           B+ Retail Inc      US bretailinc
                           Christ Chapel      US christchap
                 Prabhav Business Center   India prabhavbus
              Custom Wealth Services LLC      US customweal
Consulting Nyasa Nursing Private Limited   India consulting
                       Nexus Anchor Rain      US nexusancho
                       Moore Bitwise Inc      US moorebitwi
              Dermatology Green Medicine      US dermatolog

Number of unique Source 1 blocks: 903240
Number of unique Source 2 blocks: 1985628
Number of unique Source 3 blocks: 1941949


In [ ]:
for name, df in {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}.items():

    block_sizes = df["name_block"].value_counts()

    print(f"\n{'='*60}")
    print(name)
    print(f"{'='*60}")

    print("Average records per block:",
          round(block_sizes.mean(), 2))

    print("Median records per block:",
          block_sizes.median())

    print("Maximum records in a block:",
          block_sizes.max())

    print("Blocks containing only 1 record:",
          (block_sizes == 1).sum())

    print("Blocks containing >100 records:",
          (block_sizes > 100).sum())

    print("\nTop 10 largest blocks:")
    print(block_sizes.head(10))


Source 1
Average records per block: 2.44
Median records per block: 1.0
Maximum records in a block: 10191
Blocks containing only 1 record: 695078
Blocks containing >100 records: 1148

Top 10 largest blocks:
name_block
pediatricd    10191
innovative     5807
dermatolog     5159
physicalth     5136
cardiology     5133
earnosethr     5096
behavioral     5061
primarycar     5044
chiropract     5022
womensheal     4963
Name: count, dtype: int64

Source 2
Average records per block: 2.54
Median records per block: 1.0
Maximum records in a block: 19418
Blocks containing only 1 record: 1349702
Blocks containing >100 records: 2312

Top 10 largest blocks:
name_block
pediatricd    19418
cardiology    10302
dermatolog    10175
chiropract     9978
behavioral     9952
orthopedic     9692
physicalth     9504
primarycar     8869
internalme     8609
womensheal     8562
Name: count, dtype: int64

Source 3
Average records per block: 2.72
Median records per block: 1.0
Maximum records in a block: 243643
Bloc

In [ ]:
for name, df in {
    "Source 1": train_s1,
    "Source 2": train_s2,
    "Source 3": train_s3
}.items():

    block_sizes = df["name_block"].value_counts()

    print(f"\n{'='*60}")
    print(name)
    print(f"{'='*60}")

    print("Average records per block:",
          round(block_sizes.mean(), 2))

    print("Median records per block:",
          block_sizes.median())

    print("Maximum records in a block:",
          block_sizes.max())

    print("Blocks containing only 1 record:",
          (block_sizes == 1).sum())

    print("Blocks containing >100 records:",
          (block_sizes > 100).sum())

    print("\nTop 10 largest blocks:")
    print(block_sizes.head(10))


Source 1
Average records per block: 2.44
Median records per block: 1.0
Maximum records in a block: 10191
Blocks containing only 1 record: 695078
Blocks containing >100 records: 1148

Top 10 largest blocks:
name_block
pediatricd    10191
innovative     5807
dermatolog     5159
physicalth     5136
cardiology     5133
earnosethr     5096
behavioral     5061
primarycar     5044
chiropract     5022
womensheal     4963
Name: count, dtype: int64

Source 2
Average records per block: 2.54
Median records per block: 1.0
Maximum records in a block: 19418
Blocks containing only 1 record: 1349702
Blocks containing >100 records: 2312

Top 10 largest blocks:
name_block
pediatricd    19418
cardiology    10302
dermatolog    10175
chiropract     9978
behavioral     9952
orthopedic     9692
physicalth     9504
primarycar     8869
internalme     8609
womensheal     8562
Name: count, dtype: int64

Source 3
Average records per block: 2.72
Median records per block: 1.0
Maximum records in a block: 243643
Bloc

In [ ]:
# STEP 7 — Candidate Generation

# Create dictionaries mapping each blocking key
# to the rows belonging to that block.

s2_blocks = (
    train_s2.groupby("name_block")["entity_id"]
    .apply(list)
    .to_dict()
)

s3_blocks = (
    train_s3.groupby("name_block")["entity_id"]
    .apply(list)
    .to_dict()
)

print("Source 2 blocks:", len(s2_blocks))
print("Source 3 blocks:", len(s3_blocks))


# Generate candidates for Source 1
candidates = []

for _, row in train_s1.iterrows():

    source1_id = row["entity_id"]
    block = row["name_block"]

    # Candidates from Source 2
    for entity_id in s2_blocks.get(block, []):
        candidates.append(
            (source1_id, entity_id, "S2")
        )

    # Candidates from Source 3
    for entity_id in s3_blocks.get(block, []):
        candidates.append(
            (source1_id, entity_id, "S3")
        )


candidates_df = pd.DataFrame(
    candidates,
    columns=[
        "source1_entity_id",
        "candidate_entity_id",
        "candidate_source"
    ]
)

print("\nTotal candidate pairs:", len(candidates_df))

print("\nCandidates by source:")
print(candidates_df["candidate_source"].value_counts())

print("\nSample candidates:")
print(candidates_df.head(10))

NameError: name 'train_s2' is not defined

In [ ]:
[x for x in globals().keys() if "s1" in x.lower() or "s2" in x.lower() or "s3" in x.lower()]

[]

In [ ]:
import pandas as pd
import zipfile
import os

zip_path = "/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip"

with zipfile.ZipFile(zip_path, "r") as z:

    train_s1 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source1.tsv"),
        sep="\t"
    )

    train_s2 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source2.tsv"),
        sep="\t"
    )

    train_s3 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source3.tsv"),
        sep="\t"
    )

print("Source 1:", train_s1.shape)
print("Source 2:", train_s2.shape)
print("Source 3:", train_s3.shape)

Source 1: (2206821, 4)
Source 2: (5034616, 4)
Source 3: (5285603, 4)


In [ ]:
# STEP 7B — Fast blocking-key creation

def create_name_block(df):
    normalized = (
        df["business_name"]
        .fillna("")
        .astype(str)
        .str.lower()
        .str.replace(r"[^a-z0-9]", "", regex=True)
    )

    return normalized.str[:10]


train_s1["name_block"] = create_name_block(train_s1)
train_s2["name_block"] = create_name_block(train_s2)
train_s3["name_block"] = create_name_block(train_s3)

print("Source 1 blocking keys:")
print(train_s1[["business_name", "name_block"]].head(10))

print("\nNumber of unique Source 1 blocks:",
      train_s1["name_block"].nunique())

print("Number of unique Source 2 blocks:",
      train_s2["name_block"].nunique())

print("Number of unique Source 3 blocks:",
      train_s3["name_block"].nunique())

Source 1 blocking keys:
                              business_name  name_block
0                       Orelee's Barbershop  oreleesbar
1                               Prime Money  primemoney
2                             B+ Retail Inc  bretailinc
3                             Christ Chapel  christchap
4                   Prabhav Business Center  prabhavbus
5                Custom Wealth Services LLC  customweal
6  Consulting Nyasa Nursing Private Limited  consulting
7                         Nexus Anchor Rain  nexusancho
8                         Moore Bitwise Inc  moorebitwi
9                Dermatology Green Medicine  dermatolog

Number of unique Source 1 blocks: 903240
Number of unique Source 2 blocks: 1830053
Number of unique Source 3 blocks: 1941949


In [ ]:
import pandas as pd
import zipfile

# ============================================================
# STEP 7 — LOAD DATA + CREATE BLOCKS + GENERATE CANDIDATES
# ============================================================

zip_path = "/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip"

# ---------- 1. Load datasets ----------
with zipfile.ZipFile(zip_path, "r") as z:

    train_s1 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source1.tsv"),
        sep="\t"
    )

    train_s2 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source2.tsv"),
        sep="\t"
    )

    train_s3 = pd.read_csv(
        z.open("student_resource/dataset/train/train_source3.tsv"),
        sep="\t"
    )

print("Data loaded:")
print("Source 1:", train_s1.shape)
print("Source 2:", train_s2.shape)
print("Source 3:", train_s3.shape)


# ---------- 2. Create fast blocking keys ----------
def create_name_block(df):
    return (
        df["business_name"]
        .fillna("")
        .astype(str)
        .str.lower()
        .str.replace(r"[^a-z0-9]", "", regex=True)
        .str[:10]
    )


train_s1["name_block"] = create_name_block(train_s1)
train_s2["name_block"] = create_name_block(train_s2)
train_s3["name_block"] = create_name_block(train_s3)

print("\nBlocking keys created.")

print("Source 1 blocks:", train_s1["name_block"].nunique())
print("Source 2 blocks:", train_s2["name_block"].nunique())
print("Source 3 blocks:", train_s3["name_block"].nunique())


# ---------- 3. Generate candidates using merges ----------
s1_blocks = train_s1[
    ["entity_id", "name_block"]
].rename(
    columns={"entity_id": "source1_entity_id"}
)

s2_blocks = train_s2[
    ["entity_id", "name_block"]
].rename(
    columns={"entity_id": "candidate_entity_id"}
)

s3_blocks = train_s3[
    ["entity_id", "name_block"]
].rename(
    columns={"entity_id": "candidate_entity_id"}
)


print("\nGenerating Source 2 candidates...")

candidates_s2 = s1_blocks.merge(
    s2_blocks,
    on="name_block",
    how="inner"
)

candidates_s2["candidate_source"] = "S2"


print("Generating Source 3 candidates...")

candidates_s3 = s1_blocks.merge(
    s3_blocks,
    on="name_block",
    how="inner"
)

candidates_s3["candidate_source"] = "S3"


# ---------- 4. Combine ----------
candidates_df = pd.concat(
    [candidates_s2, candidates_s3],
    ignore_index=True
)

candidates_df = candidates_df[
    [
        "source1_entity_id",
        "candidate_entity_id",
        "candidate_source"
    ]
]


print("\n==================================================")
print("CANDIDATE GENERATION COMPLETE")
print("==================================================")

print("Total candidate pairs:", len(candidates_df))

print("\nCandidates by source:")
print(candidates_df["candidate_source"].value_counts())

print("\nSample:")
print(candidates_df.head(10))

Data loaded:
Source 1: (2206821, 4)
Source 2: (5034616, 4)
Source 3: (5285603, 4)

Blocking keys created.
Source 1 blocks: 903240
Source 2 blocks: 1830053
Source 3 blocks: 1941949

Generating Source 2 candidates...


In [2]:
print("train_s1" in globals())
print("train_s2" in globals())
print("train_s3" in globals())

True
False
False


In [8]:
import os

print("Searching your Google Drive for the dataset files...\n")

for root, dirs, files in os.walk("/content/drive/MyDrive"):
    for file in files:
        if file.lower() in [
            "train_source1.tsv",
            "train_source2.tsv",
            "train_source3.tsv",
            "train_ground_truth.tsv"
        ]:
            print(os.path.join(root, file))

Searching your Google Drive for the dataset files...



In [9]:
import glob
import os

print("Searching for ZIP files...\n")

zip_files = glob.glob(
    "/content/drive/MyDrive/**/*.zip",
    recursive=True
)

if len(zip_files) == 0:
    print("❌ No ZIP files found.")
else:
    print("ZIP files found:\n")
    for i, f in enumerate(zip_files, 1):
        print(i, "→", f)

Searching for ZIP files...

ZIP files found:

1 → /content/drive/MyDrive/bioproject/bioproject.zip
2 → /content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip


In [10]:
# ============================================================
# RECOVER DATASET + LOAD SOURCE FILES
# ============================================================

import os
import glob
import zipfile
import pandas as pd
import re
import unicodedata

print("=" * 70)
print("RECOVERING AMAZON ML CHALLENGE DATA")
print("=" * 70)

# ------------------------------------------------------------
# 1. Google Drive
# ------------------------------------------------------------

from google.colab import drive

drive.mount("/content/drive", force_remount=True)

# ------------------------------------------------------------
# 2. Locate the student resource ZIP
# ------------------------------------------------------------

zip_path = "/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip"

if not os.path.exists(zip_path):
    raise FileNotFoundError(
        f"ZIP file not found:\n{zip_path}"
    )

print("\nZIP found:")
print(zip_path)

# ------------------------------------------------------------
# 3. Inspect ZIP contents
# ------------------------------------------------------------

print("\nSearching inside ZIP...")

with zipfile.ZipFile(zip_path, "r") as z:
    all_files = z.namelist()

print(f"Total files inside ZIP: {len(all_files)}")

# Show TSV files
tsv_files = [
    f for f in all_files
    if f.lower().endswith((".tsv", ".csv"))
]

print("\nDataset files found:")
for f in tsv_files:
    print(" ->", f)

# ------------------------------------------------------------
# 4. Find Source 1, Source 2, Source 3 and Ground Truth
# ------------------------------------------------------------

def find_file(patterns):
    """
    Find a file inside the ZIP using multiple possible patterns.
    """
    for f in tsv_files:
        name = os.path.basename(f).lower()

        for pattern in patterns:
            if pattern in name:
                return f

    return None


source1_file = find_file([
    "train_source1",
    "source1"
])

source2_file = find_file([
    "train_source2",
    "source2"
])

source3_file = find_file([
    "train_source3",
    "source3"
])

ground_truth_file = find_file([
    "train_ground_truth",
    "ground_truth"
])

print("\nDetected files:")
print("Source 1      :", source1_file)
print("Source 2      :", source2_file)
print("Source 3      :", source3_file)
print("Ground Truth  :", ground_truth_file)

# ------------------------------------------------------------
# 5. Make sure required files were found
# ------------------------------------------------------------

if source1_file is None:
    raise FileNotFoundError(
        "Could not find Source 1 inside the ZIP."
    )

if source2_file is None:
    raise FileNotFoundError(
        "Could not find Source 2 inside the ZIP."
    )

if source3_file is None:
    raise FileNotFoundError(
        "Could not find Source 3 inside the ZIP."
    )

# ------------------------------------------------------------
# 6. Extract the required files
# ------------------------------------------------------------

extract_dir = "/content/amazon_ml_data"

os.makedirs(extract_dir, exist_ok=True)

with zipfile.ZipFile(zip_path, "r") as z:

    files_to_extract = [
        source1_file,
        source2_file,
        source3_file
    ]

    if ground_truth_file is not None:
        files_to_extract.append(ground_truth_file)

    for file_name in files_to_extract:
        z.extract(file_name, extract_dir)

print("\nFiles extracted successfully.")

# ------------------------------------------------------------
# 7. Get actual extracted paths
# ------------------------------------------------------------

def locate_extracted_file(filename):
    basename = os.path.basename(filename)

    matches = glob.glob(
        os.path.join(extract_dir, "**", basename),
        recursive=True
    )

    if not matches:
        raise FileNotFoundError(
            f"Could not locate extracted file: {basename}"
        )

    return matches[0]


source1_path = locate_extracted_file(source1_file)
source2_path = locate_extracted_file(source2_file)
source3_path = locate_extracted_file(source3_file)

print("\nActual paths:")
print("Source 1:", source1_path)
print("Source 2:", source2_path)
print("Source 3:", source3_path)

# ------------------------------------------------------------
# 8. Load datasets
# ------------------------------------------------------------

print("\nLoading Source 1...")
train_s1 = pd.read_csv(
    source1_path,
    sep="\t",
    low_memory=False
)

print("Source 1 shape:", train_s1.shape)

print("\nLoading Source 2...")
train_s2 = pd.read_csv(
    source2_path,
    sep="\t",
    low_memory=False
)

print("Source 2 shape:", train_s2.shape)

print("\nLoading Source 3...")
train_s3 = pd.read_csv(
    source3_path,
    sep="\t",
    low_memory=False
)

print("Source 3 shape:", train_s3.shape)

# ------------------------------------------------------------
# 9. Display columns
# ------------------------------------------------------------

print("\nSource 1 columns:")
print(train_s1.columns.tolist())

print("\nSource 2 columns:")
print(train_s2.columns.tolist())

print("\nSource 3 columns:")
print(train_s3.columns.tolist())

print("\n" + "=" * 70)
print("DATA LOADING COMPLETE")
print("=" * 70)

RECOVERING AMAZON ML CHALLENGE DATA
Mounted at /content/drive

ZIP found:
/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip

Searching inside ZIP...
Total files inside ZIP: 34

Dataset files found:
 -> student_resource/dataset/test/test_source2.tsv
 -> __MACOSX/student_resource/dataset/test/._test_source2.tsv
 -> student_resource/dataset/test/test_source3.tsv
 -> __MACOSX/student_resource/dataset/test/._test_source3.tsv
 -> student_resource/dataset/test/test_source1.tsv
 -> __MACOSX/student_resource/dataset/test/._test_source1.tsv
 -> student_resource/dataset/train/train_ground_truth.tsv
 -> __MACOSX/student_resource/dataset/train/._train_ground_truth.tsv
 -> student_resource/dataset/train/train_source2.tsv
 -> __MACOSX/student_resource/dataset/train/._train_source2.tsv
 -> student_resource/dataset/train/train_source3.tsv
 -> __MACOSX/student_resource/dataset/train/._train_source3.tsv
 -> student_resource/dataset/train/train_source1.tsv
 -> __MACOSX/student_

In [12]:
# ============================================================
# RECOVER + LOAD TRAINING DATA
# ============================================================

import os
import glob
import zipfile
import pandas as pd

print("=" * 70)
print("RECOVERING TRAINING DATA")
print("=" * 70)

# ------------------------------------------------------------
# 1. Find the Amazon ML Challenge ZIP
# ------------------------------------------------------------

zip_files = glob.glob(
    "/content/drive/MyDrive/Amazon-ML-Challenge/*.zip"
)

print("\nZIP files found:")

for f in zip_files:
    print(" ->", f)

# Find the student resource ZIP
student_zip = None

for f in zip_files:
    if "student_resource" in os.path.basename(f).lower():
        student_zip = f
        break

if student_zip is None:
    raise FileNotFoundError(
        "Amazon ML Challenge student_resource ZIP was not found."
    )

print("\nUsing ZIP:")
print(student_zip)

# ------------------------------------------------------------
# 2. Extract the TRAIN files
# ------------------------------------------------------------

extract_dir = "/content/amazon_ml_data"

os.makedirs(extract_dir, exist_ok=True)

with zipfile.ZipFile(student_zip, "r") as z:

    train_files = [
        f for f in z.namelist()
        if f.startswith("student_resource/dataset/train/")
        and not f.startswith("__MACOSX")
        and f.endswith(".tsv")
    ]

    print("\nTraining files inside ZIP:")

    for f in train_files:
        print(" ->", f)

    if len(train_files) == 0:
        raise FileNotFoundError(
            "No training TSV files were found inside the ZIP."
        )

    # Extract only training files
    for f in train_files:
        z.extract(f, extract_dir)

print("\nTraining files extracted successfully.")

# ------------------------------------------------------------
# 3. Define paths
# ------------------------------------------------------------

TRAIN_DIR = (
    "/content/amazon_ml_data/"
    "student_resource/dataset/train"
)

train_s1_path = os.path.join(
    TRAIN_DIR,
    "train_source1.tsv"
)

train_s2_path = os.path.join(
    TRAIN_DIR,
    "train_source2.tsv"
)

train_s3_path = os.path.join(
    TRAIN_DIR,
    "train_source3.tsv"
)

ground_truth_path = os.path.join(
    TRAIN_DIR,
    "train_ground_truth.tsv"
)

# ------------------------------------------------------------
# 4. Verify files actually exist
# ------------------------------------------------------------

print("\nChecking extracted files...")

required_files = {
    "Source 1": train_s1_path,
    "Source 2": train_s2_path,
    "Source 3": train_s3_path,
    "Ground Truth": ground_truth_path
}

for name, path in required_files.items():

    if os.path.exists(path):
        print("✓", name, "found")
    else:
        raise FileNotFoundError(
            f"{name} not found at:\n{path}"
        )

# ------------------------------------------------------------
# 5. Load training datasets
# ------------------------------------------------------------

print("\nLoading Source 1...")
train_s1 = pd.read_csv(
    train_s1_path,
    sep="\t"
)

print("Loading Source 2...")
train_s2 = pd.read_csv(
    train_s2_path,
    sep="\t"
)

print("Loading Source 3...")
train_s3 = pd.read_csv(
    train_s3_path,
    sep="\t"
)

print("Loading Ground Truth...")
ground_truth = pd.read_csv(
    ground_truth_path,
    sep="\t"
)

# ------------------------------------------------------------
# 6. Display results
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("TRAINING DATA LOADED SUCCESSFULLY")
print("=" * 70)

print("\nSource 1 shape:")
print(train_s1.shape)

print("\nSource 2 shape:")
print(train_s2.shape)

print("\nSource 3 shape:")
print(train_s3.shape)

print("\nGround Truth shape:")
print(ground_truth.shape)

print("\nSource 1 columns:")
print(train_s1.columns.tolist())

print("\nSource 2 columns:")
print(train_s2.columns.tolist())

print("\nSource 3 columns:")
print(train_s3.columns.tolist())

print("\nGround Truth columns:")
print(ground_truth.columns.tolist())

print("\n" + "=" * 70)
print("READY FOR STEP 7")
print("=" * 70)

RECOVERING TRAINING DATA

ZIP files found:
 -> /content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip

Using ZIP:
/content/drive/MyDrive/Amazon-ML-Challenge/6ab10eb3b23ba_student_resource.zip

Training files inside ZIP:
 -> student_resource/dataset/train/train_ground_truth.tsv
 -> student_resource/dataset/train/train_source2.tsv
 -> student_resource/dataset/train/train_source3.tsv
 -> student_resource/dataset/train/train_source1.tsv

Training files extracted successfully.

Checking extracted files...
✓ Source 1 found
✓ Source 2 found
✓ Source 3 found
✓ Ground Truth found

Loading Source 1...
Loading Source 2...
Loading Source 3...
Loading Ground Truth...

TRAINING DATA LOADED SUCCESSFULLY

Source 1 shape:
(2206821, 4)

Source 2 shape:
(5034616, 4)

Source 3 shape:
(5285603, 4)

Ground Truth shape:
(2206821, 2)

Source 1 columns:
['entity_id', 'business_name', 'business_address', 'country']

Source 2 columns:
['entity_id', 'business_name', 'business_address', 'cou

In [13]:
# ============================================================
# STEP 7 — STRONG BLOCKING + CANDIDATE ESTIMATION
# ============================================================

import pandas as pd
import re
import unicodedata

print("=" * 70)
print("STEP 7 — STRONG BLOCKING + CANDIDATE ESTIMATION")
print("=" * 70)


# ============================================================
# STEP 7A — TEXT NORMALIZATION
# ============================================================

print("\n[1/6] Creating normalized business names...")


def normalize_text(text):
    if pd.isna(text):
        return ""

    text = str(text).lower()

    # Unicode normalization
    text = unicodedata.normalize("NFKC", text)

    # Keep letters, numbers and spaces
    text = "".join(
        char if char.isalnum() or char.isspace() else " "
        for char in text
    )

    # Normalize whitespace
    text = re.sub(r"\s+", " ", text).strip()

    return text


for df in [train_s1, train_s2, train_s3]:

    if "name_normalized" not in df.columns:
        df["name_normalized"] = (
            df["business_name"]
            .fillna("")
            .astype(str)
            .apply(normalize_text)
        )


print("✓ Name normalization complete")


# ============================================================
# STEP 7B — CREATE STRONG BLOCKING KEY
# ============================================================

print("\n[2/6] Creating strong blocking keys...")


def create_strong_block(name):

    if pd.isna(name):
        return ""

    name = str(name).strip()

    if not name:
        return ""

    # Split into words
    words = name.split()

    # First word
    first_word = words[0]

    # First character of first word
    first_char = first_word[0] if first_word else ""

    # Length bucket
    length_bucket = min(len(name) // 5, 10)

    return f"{first_char}_{length_bucket}"


for df in [train_s1, train_s2, train_s3]:

    df["strong_block"] = (
        df["name_normalized"]
        .apply(create_strong_block)
    )


print("✓ Strong blocking keys created")


# ============================================================
# STEP 7C — BLOCK STATISTICS
# ============================================================

print("\n[3/6] Inspecting block statistics...")

s1_block_sizes = train_s1["strong_block"].value_counts()
s2_block_sizes = train_s2["strong_block"].value_counts()
s3_block_sizes = train_s3["strong_block"].value_counts()

print("\nSource 1 unique blocks:", len(s1_block_sizes))
print("Source 2 unique blocks:", len(s2_block_sizes))
print("Source 3 unique blocks:", len(s3_block_sizes))

print("\nLargest Source 1 blocks:")
print(s1_block_sizes.head(10))

print("\nLargest Source 2 blocks:")
print(s2_block_sizes.head(10))

print("\nLargest Source 3 blocks:")
print(s3_block_sizes.head(10))


# ============================================================
# STEP 7D — FIND COMMON BLOCKS
# ============================================================

print("\n[4/6] Finding common blocks between Source 1 and Source 2...")

common_blocks = (
    s1_block_sizes.index
    .intersection(s2_block_sizes.index)
)

print("Common blocks:", len(common_blocks))


# ============================================================
# STEP 7E — ESTIMATE CANDIDATE PAIRS
# ============================================================

print("\n[5/6] Calculating candidate count...")

s1_common = s1_block_sizes.loc[common_blocks]
s2_common = s2_block_sizes.loc[common_blocks]

candidate_counts = s1_common * s2_common

total_candidates = candidate_counts.sum()

print("\n" + "=" * 60)
print("STRONG BLOCK CANDIDATE ESTIMATE")
print("=" * 60)

print("Source 1 records:", len(train_s1))
print("Source 2 records:", len(train_s2))
print("Common blocks:", len(common_blocks))
print(
    "Estimated S1-S2 candidate pairs:",
    int(total_candidates)
)


# ============================================================
# STEP 7F — IDENTIFY LARGE BLOCKS
# ============================================================

print("\n[6/6] Inspecting potentially large blocks...")

block_summary = pd.DataFrame({
    "s1_count": s1_common,
    "s2_count": s2_common
})

block_summary["candidate_pairs"] = (
    block_summary["s1_count"]
    * block_summary["s2_count"]
)

block_summary = block_summary.sort_values(
    "candidate_pairs",
    ascending=False
)

print("\nTop 20 blocks by candidate-pair count:")

print(
    block_summary.head(20).to_string()
)


# ============================================================
# FINAL SUMMARY
# ============================================================

print("\n" + "=" * 70)
print("STEP 7 COMPLETE")
print("=" * 70)

print("\nSource 1 unique blocks:", len(s1_block_sizes))
print("Source 2 unique blocks:", len(s2_block_sizes))
print("Source 3 unique blocks:", len(s3_block_sizes))
print("Common S1-S2 blocks:", len(common_blocks))
print(
    "Estimated S1-S2 candidates:",
    f"{int(total_candidates):,}"
)

print("\nColumns now available in each source:")

print(train_s1.columns.tolist())

print("\n✓ train_s1 ready")
print("✓ train_s2 ready")
print("✓ train_s3 ready")
print("✓ Blocking analysis ready")

STEP 7 — STRONG BLOCKING + CANDIDATE ESTIMATION

[1/6] Creating normalized business names...
✓ Name normalization complete

[2/6] Creating strong blocking keys...
✓ Strong blocking keys created

[3/6] Inspecting block statistics...

Source 1 unique blocks: 375
Source 2 unique blocks: 2624
Source 3 unique blocks: 2596

Largest Source 1 blocks:
strong_block
s_5    59576
s_4    55400
s_6    50930
s_3    45240
c_4    39347
c_5    36344
a_5    34565
c_3    34561
a_4    33594
m_4    33472
Name: count, dtype: int64

Largest Source 2 blocks:
strong_block
s_4    107832
s_5    105501
s_3     91126
c_4     82442
s_6     79182
p_5     78050
c_3     74584
c_5     74390
m_4     71981
p_4     67866
Name: count, dtype: int64

Largest Source 3 blocks:
strong_block
s_4    116318
s_5    115931
s_3     96636
s_6     91320
c_4     83772
p_5     82585
c_5     76141
c_3     76068
m_4     74127
p_4     70129
Name: count, dtype: int64

[4/6] Finding common blocks between Source 1 and Source 2...
Common blocks:

# STEP 8 — EDA CONCLUSIONS

## 8.1 Dataset Overview

The training data consists of three source datasets and a ground-truth mapping:

- Source 1: 2,206,821 records
- Source 2: 5,034,616 records
- Source 3: 5,285,603 records
- Ground Truth: 2,206,821 records

All three source datasets contain the following attributes:

- `entity_id`
- `business_name`
- `business_address`
- `country`

The ground-truth dataset contains:

- `source1_entity_id`
- `matched_entity_ids`

---

## 8.2 Data Preparation

Business names were normalized before blocking.

The normalization process included:

- Converting text to lowercase
- Unicode normalization
- Removing punctuation while preserving letters and numbers
- Normalizing whitespace
- Handling missing values

A normalized business-name representation was created as `name_normalized`.

---

## 8.3 Blocking Analysis

A strong blocking key was created using the normalized business name.

The blocking analysis produced:

- Source 1 unique blocks: 375
- Source 2 unique blocks: 2,624
- Source 3 unique blocks: 2,596
- Common Source 1–Source 2 blocks: 375

Blocking is necessary because comparing every record in Source 1 with every record in Source 2 would create an extremely large comparison space.

---

## 8.4 Candidate-Pair Estimation

The current blocking strategy produces an estimated:

**93,205,788,932 S1–S2 candidate pairs**

or approximately:

**93.2 billion candidate pairs.**

The largest blocks produce billions of potential comparisons individually. For example, the `s_5` block produces approximately 6.29 billion S1–S2 candidate pairs.

Therefore, the current blocking strategy is not selective enough for direct candidate-pair generation.

---

## 8.5 Key EDA Findings

1. The three source datasets contain millions of business records, making exhaustive pairwise comparison computationally impractical.

2. Business names require normalization before similarity comparison because differences in case, punctuation, Unicode representation and whitespace can affect matching.

3. Blocking can substantially organize the search space by grouping potentially related records.

4. The current strong-blocking strategy still produces an extremely large estimated candidate space of approximately 93.2 billion S1–S2 pairs.

5. Direct materialization of all estimated candidate pairs should therefore be avoided.

6. A more selective or multi-stage blocking strategy is required before candidate generation.

---

## 8.6 Recommendations for the Next Stage

Based on the EDA and blocking analysis, the next stage should focus on:

- Refining the blocking strategy
- Using multiple blocking keys
- Reducing oversized blocks
- Generating manageable candidate sets
- Engineering similarity features for candidate pairs
- Training and evaluating entity-resolution models

These steps will be handled by the candidate-generation and machine-learning stages of the project.

---

## 8.7 EDA Lead Handoff

The EDA stage provides the downstream team with:

- Dataset structure and sizes
- Data-quality observations
- Normalization approach
- Ground-truth structure
- Blocking statistics
- Candidate-pair estimates
- Identification of the limitations of the current blocking strategy

The candidate-generation team can now use these findings to design a more selective blocking approach before generating candidate pairs.

In [14]:
# ============================================================
# STEP 8 — FINAL EDA SUMMARY
# ============================================================

print("=" * 70)
print("EDA COMPLETE — FINAL SUMMARY")
print("=" * 70)

print("\nDATASET SIZES")
print("-" * 40)
print(f"Source 1       : {len(train_s1):,}")
print(f"Source 2       : {len(train_s2):,}")
print(f"Source 3       : {len(train_s3):,}")
print(f"Ground Truth   : {len(ground_truth):,}")

print("\nBLOCKING RESULTS")
print("-" * 40)
print(f"S1 unique blocks       : {train_s1['strong_block'].nunique():,}")
print(f"S2 unique blocks       : {train_s2['strong_block'].nunique():,}")
print(f"S3 unique blocks       : {train_s3['strong_block'].nunique():,}")
print(f"Common S1-S2 blocks    : {len(common_blocks):,}")

print("\nCANDIDATE ESTIMATE")
print("-" * 40)
print(f"Estimated S1-S2 pairs  : {int(total_candidates):,}")
print(f"Approximately          : {total_candidates / 1e9:.2f} billion")

print("\n" + "=" * 70)
print("EDA LEAD WORK COMPLETE")
print("=" * 70)

print("""
Key conclusion:
The current blocking strategy produces an extremely large
candidate space. Direct candidate generation is therefore
not recommended. A more selective/multi-stage blocking
strategy is required before candidate generation.
""")

EDA COMPLETE — FINAL SUMMARY

DATASET SIZES
----------------------------------------
Source 1       : 2,206,821
Source 2       : 5,034,616
Source 3       : 5,285,603
Ground Truth   : 2,206,821

BLOCKING RESULTS
----------------------------------------
S1 unique blocks       : 375
S2 unique blocks       : 2,624
S3 unique blocks       : 2,596
Common S1-S2 blocks    : 375

CANDIDATE ESTIMATE
----------------------------------------
Estimated S1-S2 pairs  : 93,205,788,932
Approximately          : 93.21 billion

EDA LEAD WORK COMPLETE

Key conclusion:
The current blocking strategy produces an extremely large
candidate space. Direct candidate generation is therefore
not recommended. A more selective/multi-stage blocking
strategy is required before candidate generation.

